### PR TITLE
Update SharedTokenCacheCredential to filter accounts by tenant id

### LINKED ISSUE #8689: SharedTokenCacheCredential should allow users to specify the tenant in addition to username
Currently we only filter the shared token cache by `userame`, however in many cases this is not enough to distinguish the account to use. Callers should also be able to specify a tenantId as well to distinguish when their username is a guest user in multiple tenants.

### PR DESCRIPTION
**Pull request by schaabs**
- Added `SharedTokenCacheCredentialOptions` class with properties `TenantId` and `Username`
  - Added constructor overload to `SharedTokenCacheCredential` which accepts `SharedTokenCacheCredentialOptions` 
  - Added property `SharedTokenCacheTenantId` to `DefaultAzureCredentialOptions`

Fixes #8689

### CODE REVIEW

File: sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs [hunk 1]
```diff
@@ -128,53 +138,46 @@ private async ValueTask<ExtendedAccessToken> GetTokenImplAsync(TokenRequestConte
...
+                return (null, new CredentialUnavailableException(GetCredentialUnavailableMessage(accounts, filteredAccounts)));
+            }
 
-            List<IAccount> allAccounts = (await _client.GetAccountsAsync().ConfigureAwait(false)).ToList();
+            return (filteredAccounts.First(), null);
+        }
 
-            List<IAccount> accounts = (!string.IsNullOrEmpty(_username)) ? allAccounts.Where(a => a.Username == _username).ToList() : allAccounts.ToList();
+        internal const string NoAccountsInCacheMessage = "No accounts were found in the cache.  To authenticate with the SharedTokenCacheCredential, login an account through developer tooling supporting Azure single sign on.";
+        internal const string MultipleAccountsInCacheMessage = "Multiple accounts were found in the cache. To authenticate with the SharedTokenCacheCredential, set the AZURE_USERNAME and AZURE_TENANT_ID environment variables to the preferred username and tenantId, or specify them to the constructor. {0}";
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs]: Where do we read `AZURE_TENANT_ID` env var for shared token credential?

Comment (pakrym) [reply sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs]: Found it

File: sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs [hunk 2]
```diff
@@ -128,53 +138,46 @@ private async ValueTask<ExtendedAccessToken> GetTokenImplAsync(TokenRequestConte
...
+            {
+                return (null, new CredentialUnavailableException(GetCredentialUnavailableMessage(accounts, filteredAccounts)));
+            }
 
-            List<IAccount> allAccounts = (await _client.GetAccountsAsync().ConfigureAwait(false)).ToList();
+            return (filteredAccounts.First(), null);
+        }
 
-            List<IAccount> accounts = (!string.IsNullOrEmpty(_username)) ? allAccounts.Where(a => a.Username == _username).ToList() : allAccounts.ToList();
+        internal const string NoAccountsInCacheMessage = "No accounts were found in the cache.  To authenticate with the SharedTokenCacheCredential, login an account through developer tooling supporting Azure single sign on.";
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs]: nit: Move these message templates to the top of the class

File: sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs [hunk 3]
```diff
@@ -128,53 +138,46 @@ private async ValueTask<ExtendedAccessToken> GetTokenImplAsync(TokenRequestConte
...
-                }
-                else
-                {
-                    ex = new CredentialUnavailableException($"User account '{_username}' was not found in the shared token cache.{Environment.NewLine}  Discovered Accounts: [ '{string.Join("', '", allAccounts.Select(a => a.Username))}' ]");
-                }
+                return NoAccountsInCacheMessage;
             }
-            else
+
+            var discoveredAccountsStr = $"[ {{{string.Join("}, {", accounts.Select(a => $"username: {a.Username} tenantId: {a.HomeAccountId?.TenantId}"))}}} ]";
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs]: What about joining them with a newline instead of `[ {}, {} ]`?

Comment (schaabs) [reply sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs]: I tried the newline but the formatting looked a bit strange as it didn't respect the indentations when printed out in Exception.ToString().

File: sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs (hunk 1 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs]: In `DefaultAzureCredential`. Currently both `AZURE_USERNAME` and `AZURE_TENANT_ID` are only respected when instantiating the `SharedTokenCacheCredential` through the `DefaultAzureCredential`. I'm on the fence as to whether to change this behavior. Currently only the `DefaultAzureCredential` and the `EnvironmentCredential` actually read environment variables and the rest of the credential types need to be explicitly passed values for these. It's possible to change this so the `SharedTokenCacheCredential` will read the environment variable itself, but then perhaps all other credential types such as `ClientSecretCredential`, `ClientCertificateCredential`, `InteractiveBrowserCredential` etc. would also directly respect these environment variables.

I'm not opposed to changing this behavior throughout the library, but if we do, it should be done as part of a larger work item.

**Review by pakrym — APPROVED** [review]

File: sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs (hunk 1 above)
Comment (pakrym) [reply sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs]: File an issue to decide this please. We need to do it soon/

Comment (schaabs) [reply sdk/identity/Azure.Identity/src/SharedTokenCacheCredential.cs]: Added issue https://github.com/Azure/azure-sdk-for-net/issues/8692 to track this.

**Comment by pakrym** [comment]
You didn't move the messages :(

**Pull request merged**